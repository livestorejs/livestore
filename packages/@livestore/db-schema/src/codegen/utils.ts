import os from 'node:os'
import path from 'node:path'

import dprint from 'dprint-node'

export const pretty = (code: string): string => {
  const tmpPath = os.tmpdir()
  const filePath = path.join(tmpPath, 'tmp.ts')

  return dprint.format(filePath, code, {
    semiColons: 'asi',
    quoteStyle: 'preferSingle',
  })
}
