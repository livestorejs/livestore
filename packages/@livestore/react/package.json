{
  "name": "@livestore/react",
  "version": "0.3.0-dev.38",
  "type": "module",
  "sideEffects": false,
  "exports": {
    ".": {
      "types": "./dist/mod.d.ts",
      "default": "./dist/mod.js"
    },
    "./experimental": {
      "types": "./dist/experimental/mod.d.ts",
      "default": "./dist/experimental/mod.js"
    }
  },
  "types": "./dist/mod.d.ts",
  "typesVersions": {
    "*": {
      "./experimental": [
        "./dist/experimental/mod.d.ts"
      ]
    }
  },
  "dependencies": {
    "@livestore/common": "workspace:*",
    "@livestore/livestore": "workspace:*",
    "@livestore/utils": "workspace:*",
    "@opentelemetry/api": "1.9.0"
  },
  "devDependencies": {
    "@livestore/adapter-web": "workspace:*",
    "@livestore/utils-dev": "workspace:*",
    "@opentelemetry/sdk-trace-base": "^2.0.0",
    "@testing-library/dom": "^10.4.0",
    "@testing-library/react": "^16.2.0",
    "@types/react": "^19.1.2",
    "@types/react-dom": "^19.0.4",
    "jsdom": "^26.0.0",
    "react": "^19.0.0",
    "react-dom": "^19.0.0",
    "react-window": "^1.8.11",
    "typescript": "^5.8.3",
    "vite": "^6.2.1",
    "vitest": "^3.1.1"
  },
  "files": [
    "package.json",
    "src",
    "dist"
  ],
  "peerDependencies": {
    "react": "~19.0.0"
  },
  "publishConfig": {
    "access": "public"
  },
  "scripts": {
    "build": "tsc",
    "test": "vitest && REACT_STRICT_MODE=1 vitest"
  }
}
