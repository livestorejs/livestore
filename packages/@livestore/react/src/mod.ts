export { LiveStoreContext, type ReactApi } from './LiveStoreContext.js'
export { useStore, withReactApi } from './useStore.js'
export { LiveStoreProvider } from './LiveStoreProvider.js'
export { useStackInfo } from './utils/stack-info.js'
export { useQuery, useQueryRef } from './useQuery.js'
export {
  useClientDocument,
  type StateSetters,
  type SetStateAction,
  type Dispatch,
  type UseRowResult as UseStateResult,
} from './useClientDocument.js'
