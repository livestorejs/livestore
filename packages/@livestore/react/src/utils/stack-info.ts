import { extractStackInfoFromStackTrace, type StackInfo } from '@livestore/livestore'
import React from 'react'

export const originalStackLimit = Error.stackTraceLimit

export const useStackInfo = (): StackInfo =>
  React.useMemo(() => {
    Error.stackTraceLimit = 10
    // eslint-disable-next-line unicorn/error-message
    const stack = new Error().stack!
    Error.stackTraceLimit = originalStackLimit
    return extractStackInfoFromStackTrace(stack)
  }, [])
