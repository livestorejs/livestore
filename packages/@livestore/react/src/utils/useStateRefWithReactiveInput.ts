import React from 'react'

/**
 * A variant of `React.useState` which allows the `inputState` to change over time as well.
 * Important: This hook is synchronous / single-render-pass (i.e. doesn't use `useEffect` or `setState` directly).
 *
 * Notes:
 * - The output state is always reset to the input state in case the input state changes (i.e. the previous "external" `setStateAndRerender` call is forgotten)
 * - This hook might not work properly with React Suspense
 * - Also see this Tweet for more potential problems: https://twitter.com/schickling/status/1677317711104278528
 *
 */
export const useStateRefWithReactiveInput = <T>(
  inputState: T,
): [React.MutableRefObject<T>, (newState: T | ((prev: T) => T)) => void] => {
  const [_, rerender] = React.useState(0)

  const lastKnownInputStateRef = React.useRef<T>(inputState)
  const stateRef = React.useRef<T>(inputState)

  if (lastKnownInputStateRef.current !== inputState) {
    lastKnownInputStateRef.current = inputState

    // NOTE we don't need to re-render here, because the component is already re-rendering due to the `inputState` change
    stateRef.current = inputState
  }

  const setStateAndRerender = React.useCallback(
    (newState: ((prev: T) => T) | T) => {
      // @ts-expect-error https://github.com/microsoft/TypeScript/issues/37663
      const val = typeof newState === 'function' ? newState(stateRef.current) : newState
      stateRef.current = val
      rerender((c) => c + 1)
    },
    [rerender],
  )

  return [stateRef, setStateAndRerender]
}

// Down-side of this implementation: Double render pass due to `setState` call (which forces a re-render)
// Keeping around for now in case `useStateRefWithReactiveInput` doesn't work out
// const _useStateWithReactiveInput = <T>(inputState: T): [T, (newState: T | ((prev: T) => T)) => void] => {
//   const [externalState, setExternalState] = React.useState(inputState)

//   if (externalState !== inputState) {
//     setExternalState(inputState)
//   }

//   return [externalState, setExternalState]
// }
