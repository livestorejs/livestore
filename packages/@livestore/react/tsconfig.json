{
  "extends": "../../../tsconfig.base.json",
  "compilerOptions": {
    "outDir": "./dist",
    "rootDir": "./src",
    "jsx": "react",
    "skipLibCheck": true,
    "resolveJsonModule": true,
    // "jsx": "preserve",
    "tsBuildInfoFile": "./dist/.tsbuildinfo"
  },
  "include": ["./src"],
  "references": [
    { "path": "../db-schema" },
    { "path": "../common" },
    { "path": "../web" },
    { "path": "../livestore" },
    { "path": "../utils" }
  ]
}
