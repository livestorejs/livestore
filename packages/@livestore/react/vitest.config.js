import { defineConfig } from 'vite'

export default defineConfig({
  test: {
    // Needed for React hook tests
    environment: 'jsdom',
  },
  esbuild: {
    // TODO remove once `using` keyword supported OOTB with Vite https://github.com/vitejs/vite/issues/15464#issuecomment-1872485703
    target: 'es2020',
  },
  resolve: {
    alias: {
      '@livestore/wa-sqlite/dist/wa-sqlite.mjs': '@livestore/wa-sqlite/dist/wa-sqlite.node.mjs',
    },
  },
})
