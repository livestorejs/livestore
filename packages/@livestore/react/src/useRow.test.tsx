import * as LiveStore from '@livestore/livestore'
import { getSimplifiedRootSpan } from '@livestore/livestore/internal/testing-utils'
import { Effect, ReadonlyRecord, Schema } from '@livestore/utils/effect'
import * as otel from '@opentelemetry/api'
import { BasicTracerProvider, InMemorySpanExporter, SimpleSpanProcessor } from '@opentelemetry/sdk-trace-base'
import { render, renderHook } from '@testing-library/react'
import React from 'react'
import { describe, expect, it } from 'vitest'

import { AppComponentSchema, AppRouterSchema, makeTodoMvcReact, tables, todos } from './__tests__/fixture.js'
import * as LiveStoreReact from './mod.js'

// NOTE running tests concurrently doesn't work with the default global db graph
describe('useRow', () => {
  it('should update the data based on component key', () =>
    Effect.gen(function* () {
      const { wrapper, store, reactivityGraph, makeRenderCount } = yield* makeTodoMvcReact({
        useGlobalReactivityGraph: false,
      })

      const renderCount = makeRenderCount()

      const { result, rerender } = renderHook(
        (userId: string) => {
          renderCount.inc()

          const [state, setState] = LiveStoreReact.useRow(AppComponentSchema, userId, { reactivityGraph })
          return { state, setState }
        },
        { wrapper, initialProps: 'u1' },
      )

      expect(result.current.state.id).toBe('u1')
      expect(result.current.state.username).toBe('')
      expect(renderCount.val).toBe(1)

      React.act(() =>
        store.mutate(
          LiveStore.rawSqlMutation({
            sql: LiveStore.sql`INSERT INTO UserInfo (id, username) VALUES ('u2', 'username_u2')`,
          }),
        ),
      )

      rerender('u2')

      expect(result.current.state.id).toBe('u2')
      expect(result.current.state.username).toBe('username_u2')
      expect(renderCount.val).toBe(2)
    }).pipe(Effect.scoped, Effect.tapCauseLogPretty, Effect.runPromise))

  // TODO add a test that makes sure React doesn't re-render when a setter is used to set the same value

  it('should update the data reactively - via setState', () =>
    Effect.gen(function* () {
      const { wrapper, reactivityGraph, makeRenderCount } = yield* makeTodoMvcReact({
        useGlobalReactivityGraph: false,
      })

      const renderCount = makeRenderCount()

      const { result } = renderHook(
        (userId: string) => {
          renderCount.inc()

          const [state, setState] = LiveStoreReact.useRow(AppComponentSchema, userId, { reactivityGraph })
          return { state, setState }
        },
        { wrapper, initialProps: 'u1' },
      )

      expect(result.current.state.id).toBe('u1')
      expect(result.current.state.username).toBe('')
      expect(renderCount.val).toBe(1)

      React.act(() => result.current.setState.username('username_u1_hello'))

      expect(result.current.state.id).toBe('u1')
      expect(result.current.state.username).toBe('username_u1_hello')
      expect(renderCount.val).toBe(2)
    }).pipe(Effect.scoped, Effect.tapCauseLogPretty, Effect.runPromise))

  it('should update the data reactively - via raw store mutation', () =>
    Effect.gen(function* () {
      const { wrapper, store, reactivityGraph, makeRenderCount } = yield* makeTodoMvcReact({
        useGlobalReactivityGraph: false,
      })

      const renderCount = makeRenderCount()

      const { result } = renderHook(
        (userId: string) => {
          renderCount.inc()

          const [state, setState] = LiveStoreReact.useRow(AppComponentSchema, userId, { reactivityGraph })
          return { state, setState }
        },
        { wrapper, initialProps: 'u1' },
      )

      expect(result.current.state.id).toBe('u1')
      expect(result.current.state.username).toBe('')
      expect(renderCount.val).toBe(1)

      React.act(() =>
        store.mutate(
          LiveStore.rawSqlMutation({
            sql: LiveStore.sql`UPDATE UserInfo SET username = 'username_u1_hello' WHERE id = 'u1';`,
          }),
        ),
      )

      expect(result.current.state.id).toBe('u1')
      expect(result.current.state.username).toBe('username_u1_hello')
      expect(renderCount.val).toBe(2)
    }).pipe(Effect.scoped, Effect.tapCauseLogPretty, Effect.runPromise))

  it('should work for a larger app', () =>
    Effect.gen(function* () {
      const { wrapper, store, reactivityGraph, makeRenderCount } = yield* makeTodoMvcReact({
        useGlobalReactivityGraph: false,
      })

      const allTodos$ = LiveStore.querySQL(`select * from todos`, {
        label: 'allTodos',
        schema: Schema.Array(tables.todos.schema),
        reactivityGraph,
      })

      const appRouterRenderCount = makeRenderCount()
      let globalSetState: LiveStoreReact.StateSetters<typeof AppRouterSchema> | undefined
      const AppRouter: React.FC = () => {
        appRouterRenderCount.inc()

        const [state, setState] = LiveStoreReact.useRow(AppRouterSchema, { reactivityGraph })

        globalSetState = setState

        return (
          <div>
            <TasksList setTaskId={setState.currentTaskId} />
            <div role="current-id">Current Task Id: {state.currentTaskId ?? '-'}</div>
            {state.currentTaskId ? <TaskDetails id={state.currentTaskId} /> : <div>Click on a task to see details</div>}
          </div>
        )
      }

      const TasksList: React.FC<{ setTaskId: (_: string) => void }> = ({ setTaskId }) => {
        const allTodos = LiveStoreReact.useQuery(allTodos$)

        return (
          <div>
            {allTodos.map((_) => (
              <div key={_.id} onClick={() => setTaskId(_.id)}>
                {_.id}
              </div>
            ))}
          </div>
        )
      }

      const TaskDetails: React.FC<{ id: string }> = ({ id }) => {
        const [todo] = LiveStoreReact.useRow(todos, id, { reactivityGraph })
        return <div role="content">{JSON.stringify(todo)}</div>
      }

      const renderResult = render(<AppRouter />, { wrapper })

      expect(appRouterRenderCount.val).toBe(1)

      React.act(() =>
        store.mutate(
          LiveStore.rawSqlMutation({
            sql: LiveStore.sql`INSERT INTO todos (id, text, completed) VALUES ('t1', 'buy milk', 0)`,
          }),
        ),
      )

      expect(appRouterRenderCount.val).toBe(1)
      expect(renderResult.getByRole('current-id').innerHTML).toMatchInlineSnapshot('"Current Task Id: -"')

      React.act(() => globalSetState!.currentTaskId('t1'))

      expect(appRouterRenderCount.val).toBe(2)
      expect(renderResult.getByRole('content').innerHTML).toMatchInlineSnapshot(
        `"{"id":"t1","text":"buy milk","completed":false}"`,
      )

      expect(renderResult.getByRole('current-id').innerHTML).toMatchInlineSnapshot('"Current Task Id: t1"')

      React.act(() =>
        store.mutate(
          LiveStore.rawSqlMutation({
            sql: LiveStore.sql`INSERT INTO todos (id, text, completed) VALUES ('t2', 'buy eggs', 0)`,
          }),
          AppRouterSchema.update({ where: { id: 'singleton' }, values: { currentTaskId: 't2' } }),
          LiveStore.rawSqlMutation({
            sql: LiveStore.sql`INSERT INTO todos (id, text, completed) VALUES ('t3', 'buy bread', 0)`,
          }),
        ),
      )

      expect(appRouterRenderCount.val).toBe(3)
      expect(renderResult.getByRole('current-id').innerHTML).toMatchInlineSnapshot('"Current Task Id: t2"')
    }).pipe(Effect.scoped, Effect.tapCauseLogPretty, Effect.runPromise))

  it('should work for a useRow query chained with a useTemporary query', () =>
    Effect.gen(function* () {
      const { store, wrapper, reactivityGraph, makeRenderCount } = yield* makeTodoMvcReact({
        useGlobalReactivityGraph: false,
      })
      const renderCount = makeRenderCount()

      store.mutate(
        todos.insert({ id: 't1', text: 'buy milk', completed: false }),
        todos.insert({ id: 't2', text: 'buy bread', completed: false }),
      )

      const { result, unmount, rerender } = renderHook(
        (userId: string) => {
          renderCount.inc()

          const [_row, _setRow, rowState$] = LiveStoreReact.useRow(AppComponentSchema, userId, { reactivityGraph })
          const todos = LiveStoreReact.useScopedQuery(
            () =>
              LiveStore.querySQL(
                (get) => LiveStore.sql`select * from todos where text like '%${get(rowState$).text}%'`,
                {
                  schema: Schema.Array(tables.todos.schema),
                  reactivityGraph,
                  label: 'todosFiltered',
                },
              ),
            userId,
          )

          return { todos }
        },
        { wrapper, initialProps: 'u1' },
      )

      React.act(() =>
        store.mutate(
          LiveStore.rawSqlMutation({
            sql: LiveStore.sql`INSERT INTO UserInfo (id, username, text) VALUES ('u2', 'username_u2', 'milk')`,
          }),
        ),
      )

      expect(result.current.todos.length).toBe(2)
      // expect(result.current.state.username).toBe('')
      expect(renderCount.val).toBe(1)

      rerender('u2')

      expect(result.current.todos.length).toBe(1)
      expect(renderCount.val).toBe(2)

      unmount()
    }).pipe(Effect.scoped, Effect.tapCauseLogPretty, Effect.runPromise))

  let cachedProvider: BasicTracerProvider | undefined

  describe('otel', () => {
    const exporter = new InMemorySpanExporter()

    const provider = cachedProvider ?? new BasicTracerProvider()
    cachedProvider = provider
    provider.addSpanProcessor(new SimpleSpanProcessor(exporter))
    provider.register()

    const otelTracer = otel.trace.getTracer('test')

    const span = otelTracer.startSpan('test')
    const otelContext = otel.trace.setSpan(otel.context.active(), span)

    it('should update the data based on component key', async () => {
      const { strictMode } = await Effect.gen(function* () {
        const { wrapper, store, reactivityGraph, makeRenderCount, strictMode } = yield* makeTodoMvcReact({
          useGlobalReactivityGraph: false,
          otelContext,
          otelTracer,
        })

        const renderCount = makeRenderCount()

        const { result, rerender, unmount } = renderHook(
          (userId: string) => {
            renderCount.inc()

            const [state, setState] = LiveStoreReact.useRow(AppComponentSchema, userId, { reactivityGraph })
            return { state, setState }
          },
          { wrapper, initialProps: 'u1' },
        )

        expect(result.current.state.id).toBe('u1')
        expect(result.current.state.username).toBe('')
        expect(renderCount.val).toBe(1)

        React.act(() =>
          store.mutate(
            LiveStore.rawSqlMutation({
              sql: LiveStore.sql`INSERT INTO UserInfo (id, username) VALUES ('u2', 'username_u2')`,
            }),
          ),
        )

        rerender('u2')

        expect(result.current.state.id).toBe('u2')
        expect(result.current.state.username).toBe('username_u2')
        expect(renderCount.val).toBe(2)

        unmount()
        span.end()

        return { strictMode }
      }).pipe(Effect.scoped, Effect.tapCauseLogPretty, Effect.runPromise)

      const mapAttributes = (attributes: otel.Attributes) => {
        return ReadonlyRecord.map(attributes, (val, key) => {
          if (key === 'stackInfo') {
            const stackInfo = JSON.parse(val as string) as LiveStore.StackInfo
            // stackInfo.frames.shift() // Removes `renderHook.wrapper` from the stack
            stackInfo.frames.forEach((_) => {
              if (_.name.includes('renderHook.wrapper')) {
                _.name = 'renderHook.wrapper'
              }
              _.filePath = '__REPLACED_FOR_SNAPSHOT__'
            })
            return JSON.stringify(stackInfo)
          }
          return val
        })
      }

      // TODO improve testing setup so "obsolete" warning is avoided
      if (strictMode) {
        expect(getSimplifiedRootSpan(exporter, mapAttributes)).toMatchSnapshot('strictMode=true')
      } else {
        expect(getSimplifiedRootSpan(exporter, mapAttributes)).toMatchSnapshot('strictMode=false')
      }
    })
  })
})
