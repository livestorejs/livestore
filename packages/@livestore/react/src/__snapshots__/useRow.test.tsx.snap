// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useRow > otel > should update the data based on component key > strictMode=false 1`] = `
{
  "_name": "test",
  "children": [
    {
      "_name": "livestore.in-memory-db:execute",
      "attributes": {
        "sql.query": "
      PRAGMA page_size=32768;
      PRAGMA cache_size=10000;
      PRAGMA journal_mode='MEMORY'; -- we don't flush to disk before committing a write
      PRAGMA synchronous='OFF';
      PRAGMA temp_store='MEMORY';
      PRAGMA foreign_keys='ON'; -- we want foreign key constraints to be enforced
    ",
      },
    },
    {
      "_name": "sql-in-memory-select",
      "attributes": {
        "sql.cached": false,
        "sql.query": "SELECT 1 FROM 'UserInfo' WHERE id = ?",
        "sql.rowsCount": 0,
      },
    },
    {
      "_name": "sql-in-memory-select",
      "attributes": {
        "sql.cached": false,
        "sql.query": "SELECT 1 FROM 'UserInfo' WHERE id = ?",
        "sql.rowsCount": 1,
      },
    },
    {
      "_name": "LiveStore:mutations",
      "children": [
        {
          "_name": "LiveStore:mutate",
          "attributes": {
            "livestore.mutateLabel": "mutate",
          },
          "children": [
            {
              "_name": "LiveStore:processWrites",
              "attributes": {
                "livestore.mutateLabel": "mutate",
              },
              "children": [
                {
                  "_name": "LiveStore:mutateWithoutRefresh",
                  "attributes": {
                    "livestore.args": "{
  "sql": "INSERT INTO UserInfo (id, username) VALUES ('u2', 'username_u2')"
}",
                    "livestore.mutation": "livestore.RawSql",
                  },
                  "children": [
                    {
                      "_name": "livestore.in-memory-db:execute",
                      "attributes": {
                        "sql.query": "INSERT INTO UserInfo (id, username) VALUES ('u2', 'username_u2')",
                      },
                    },
                  ],
                },
              ],
            },
          ],
        },
      ],
    },
    {
      "_name": "LiveStore:queries",
      "children": [
        {
          "_name": "db:SELECT * FROM 'UserInfo' WHERE id = ?",
          "attributes": {
            "sql.query": "SELECT * FROM 'UserInfo' WHERE id = ?",
            "sql.rowsCount": 1,
          },
          "children": [
            {
              "_name": "sql-in-memory-select",
              "attributes": {
                "sql.cached": false,
                "sql.query": "SELECT * FROM 'UserInfo' WHERE id = ?",
                "sql.rowsCount": 1,
              },
            },
          ],
        },
        {
          "_name": "LiveStore:useRow:UserInfo:u1",
          "attributes": {
            "id": "u1",
          },
          "children": [
            {
              "_name": "LiveStore:mutateWithoutRefresh",
              "attributes": {
                "livestore.args": "{
  "id": "u1"
}",
                "livestore.mutation": "_Derived_Create_UserInfo",
              },
              "children": [
                {
                  "_name": "livestore.in-memory-db:execute",
                  "attributes": {
                    "sql.query": "INSERT INTO UserInfo (username, text, id) VALUES ($username, $text, $id)",
                  },
                },
              ],
            },
            {
              "_name": "LiveStore:useQuery:db(row:UserInfo:u1)",
              "attributes": {
                "label": "db(row:UserInfo:u1)",
                "stackInfo": "{"frames":[{"name":"renderHook.wrapper","filePath":"__REPLACED_FOR_SNAPSHOT__"},{"name":"useRow","filePath":"__REPLACED_FOR_SNAPSHOT__"}]}",
              },
              "children": [
                {
                  "_name": "db:SELECT * FROM 'UserInfo' WHERE id = ?",
                  "attributes": {
                    "sql.query": "SELECT * FROM 'UserInfo' WHERE id = ?",
                    "sql.rowsCount": 1,
                  },
                  "children": [
                    {
                      "_name": "sql-in-memory-select",
                      "attributes": {
                        "sql.cached": false,
                        "sql.query": "SELECT * FROM 'UserInfo' WHERE id = ?",
                        "sql.rowsCount": 1,
                      },
                    },
                  ],
                },
                {
                  "_name": "LiveStore.subscribe",
                  "attributes": {
                    "label": "db(row:UserInfo:u1)",
                    "queryLabel": "db(row:UserInfo:u1)",
                  },
                },
              ],
            },
          ],
        },
        {
          "_name": "LiveStore:useRow:UserInfo:u2",
          "attributes": {
            "id": "u2",
          },
          "children": [
            {
              "_name": "LiveStore:useQuery:db(row:UserInfo:u2)",
              "attributes": {
                "label": "db(row:UserInfo:u2)",
                "stackInfo": "{"frames":[{"name":"renderHook.wrapper","filePath":"__REPLACED_FOR_SNAPSHOT__"},{"name":"useRow","filePath":"__REPLACED_FOR_SNAPSHOT__"}]}",
              },
              "children": [
                {
                  "_name": "db:SELECT * FROM 'UserInfo' WHERE id = ?",
                  "attributes": {
                    "sql.query": "SELECT * FROM 'UserInfo' WHERE id = ?",
                    "sql.rowsCount": 1,
                  },
                  "children": [
                    {
                      "_name": "sql-in-memory-select",
                      "attributes": {
                        "sql.cached": false,
                        "sql.query": "SELECT * FROM 'UserInfo' WHERE id = ?",
                        "sql.rowsCount": 1,
                      },
                    },
                  ],
                },
                {
                  "_name": "LiveStore.subscribe",
                  "attributes": {
                    "label": "db(row:UserInfo:u2)",
                    "queryLabel": "db(row:UserInfo:u2)",
                  },
                },
              ],
            },
          ],
        },
      ],
    },
  ],
}
`;

exports[`useRow > otel > should update the data based on component key > strictMode=true 1`] = `
{
  "_name": "test",
  "children": [
    {
      "_name": "livestore.in-memory-db:execute",
      "attributes": {
        "sql.query": "
      PRAGMA page_size=32768;
      PRAGMA cache_size=10000;
      PRAGMA journal_mode='MEMORY'; -- we don't flush to disk before committing a write
      PRAGMA synchronous='OFF';
      PRAGMA temp_store='MEMORY';
      PRAGMA foreign_keys='ON'; -- we want foreign key constraints to be enforced
    ",
      },
    },
    {
      "_name": "sql-in-memory-select",
      "attributes": {
        "sql.cached": false,
        "sql.query": "SELECT 1 FROM 'UserInfo' WHERE id = ?",
        "sql.rowsCount": 0,
      },
    },
    {
      "_name": "sql-in-memory-select",
      "attributes": {
        "sql.cached": false,
        "sql.query": "SELECT 1 FROM 'UserInfo' WHERE id = ?",
        "sql.rowsCount": 1,
      },
    },
    {
      "_name": "LiveStore:mutations",
      "children": [
        {
          "_name": "LiveStore:mutate",
          "attributes": {
            "livestore.mutateLabel": "mutate",
          },
          "children": [
            {
              "_name": "LiveStore:processWrites",
              "attributes": {
                "livestore.mutateLabel": "mutate",
              },
              "children": [
                {
                  "_name": "LiveStore:mutateWithoutRefresh",
                  "attributes": {
                    "livestore.args": "{
  "sql": "INSERT INTO UserInfo (id, username) VALUES ('u2', 'username_u2')"
}",
                    "livestore.mutation": "livestore.RawSql",
                  },
                  "children": [
                    {
                      "_name": "livestore.in-memory-db:execute",
                      "attributes": {
                        "sql.query": "INSERT INTO UserInfo (id, username) VALUES ('u2', 'username_u2')",
                      },
                    },
                  ],
                },
              ],
            },
          ],
        },
      ],
    },
    {
      "_name": "LiveStore:queries",
      "children": [
        {
          "_name": "db:SELECT * FROM 'UserInfo' WHERE id = ?",
          "attributes": {
            "sql.query": "SELECT * FROM 'UserInfo' WHERE id = ?",
            "sql.rowsCount": 1,
          },
          "children": [
            {
              "_name": "sql-in-memory-select",
              "attributes": {
                "sql.cached": false,
                "sql.query": "SELECT * FROM 'UserInfo' WHERE id = ?",
                "sql.rowsCount": 1,
              },
            },
          ],
        },
        {
          "_name": "LiveStore:useRow:UserInfo:u1",
          "attributes": {
            "id": "u1",
          },
          "children": [
            {
              "_name": "LiveStore:mutateWithoutRefresh",
              "attributes": {
                "livestore.args": "{
  "id": "u1"
}",
                "livestore.mutation": "_Derived_Create_UserInfo",
              },
              "children": [
                {
                  "_name": "livestore.in-memory-db:execute",
                  "attributes": {
                    "sql.query": "INSERT INTO UserInfo (username, text, id) VALUES ($username, $text, $id)",
                  },
                },
              ],
            },
            {
              "_name": "LiveStore:useQuery:db(row:UserInfo:u1)",
              "attributes": {
                "label": "db(row:UserInfo:u1)",
                "stackInfo": "{"frames":[{"name":"renderHook.wrapper","filePath":"__REPLACED_FOR_SNAPSHOT__"},{"name":"useRow","filePath":"__REPLACED_FOR_SNAPSHOT__"}]}",
              },
              "children": [
                {
                  "_name": "db:SELECT * FROM 'UserInfo' WHERE id = ?",
                  "attributes": {
                    "sql.query": "SELECT * FROM 'UserInfo' WHERE id = ?",
                    "sql.rowsCount": 1,
                  },
                  "children": [
                    {
                      "_name": "sql-in-memory-select",
                      "attributes": {
                        "sql.cached": false,
                        "sql.query": "SELECT * FROM 'UserInfo' WHERE id = ?",
                        "sql.rowsCount": 1,
                      },
                    },
                  ],
                },
                {
                  "_name": "LiveStore.subscribe",
                  "attributes": {
                    "label": "db(row:UserInfo:u1)",
                    "queryLabel": "db(row:UserInfo:u1)",
                  },
                },
                {
                  "_name": "LiveStore.subscribe",
                  "attributes": {
                    "label": "db(row:UserInfo:u1)",
                    "queryLabel": "db(row:UserInfo:u1)",
                  },
                },
              ],
            },
          ],
        },
      ],
    },
  ],
}
`;
